### PR TITLE
fix #27933, don't `eval` value of global vars in cluster deserializer

### LINKED ISSUE #27933: Global `Symbol` and `Expr` get interpreted when copying to remote
This happens in both 0.6.3 and 0.7-beta on Windows.

Transfering `String` and local `Symbol` are fine:

```
julia> addprocs(1)
1-element Array{Int64,1}:
 2

julia> a = "foo"
"foo"

julia> remotecall_fetch(()->println(a), 2)
      From worker 2:    foo

julia> let a = :foo
              remotecall_fetch(()->println(a), 2)
          end
      From worker 2:    foo

```

However, global symbol get interpreted in worker:

```
julia> a = :foo
:foo

julia> remotecall_fetch(()->println(a), 2)
ERROR: On worker 2:
UndefVarError: foo not defined
top-level scope at none:0
eval at .\boot.jl:319 [inlined]
deserialize_global_from_main at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\clusterserialize.jl:178
#5 at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\clusterserialize.jl:72 [inlined]
foreach at .\abstractarray.jl:1819
deserialize at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\clusterserialize.jl:72
handle_deserialize at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Serialization\src\Serialization.jl:800
deserialize at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Serialization\src\Serialization.jl:706
handle_deserialize at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Serialization\src\Serialization.jl:750
deserialize at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Serialization\src\Serialization.jl:706
handle_deserialize at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Serialization\src\Serialization.jl:753
deserialize_msg at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Serialization\src\Serialization.jl:706
#invokelatest#1 at .\essentials.jl:670 [inlined]
invokelatest at .\essentials.jl:669 [inlined]
message_handler_loop at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\process_messages.jl:160
process_tcp_streams at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\process_messages.jl:117
#105 at .\task.jl:257
Stacktrace:
 [1] #remotecall_fetch#149 at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\remotecall.jl:379 [inlined]
 [2] remotecall_fetch(::Function, ::Distributed.Worker) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\remotecall.jl:371
 [3] #remotecall_fetch#152(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Function, ::Int64) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\remotecall.jl:392
 [4] remotecall_fetch(::Function, ::Int64) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Distributed\src\remotecall.jl:392
 [5] top-level scope at none:0
```

The same happens for `Expr`s

```
julia> a = :(1+1)
:(1 + 1)

julia> println(a)
1 + 1

julia> remotecall_fetch(()->println(a), 2)
      From worker 2:    2
```

Following the stacktrace I think the problem is that we [eval the value when deserializing](https://github.com/JuliaLang/julia/blob/master/stdlib/Distributed/src/clusterserialize.jl#L178). We should probably add special checks for `Symbol` and `Expr` since they behave differently when being evaled.

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

File: stdlib/Distributed/src/clusterserialize.jl [hunk 1]
```diff
@@ -173,9 +173,9 @@ function deserialize_global_from_main(s::ClusterSerializer, sym)
     sym_isconst = deserialize(s)
     v = deserialize(s)
     if sym_isconst
-        @eval Main const $sym = $v
+        ccall(:jl_set_const, Cvoid, (Any, Any, Any), Main, sym, v)
     else
-        @eval Main $sym = $v
+        ccall(:jl_set_global, Cvoid, (Any, Any, Any), Main, sym, v)
     end
 end
```
Comment (vtjnash) [inline stdlib/Distributed/src/clusterserialize.jl]: This changes the implicit value of the function. Does that matter? Should we still have `return v` at the end?

File: stdlib/Distributed/test/distributed_exec.jl [hunk 2]
```diff
@@ -1501,6 +1501,10 @@ else
     @info "SO_REUSEPORT is unsupported, skipping reuseport tests"
 end
 
+# issue #27933
+a27933 = :_not_defined_27933
+@test remotecall_fetch(()->a, 2) === a27933
```
Comment (Keno) [inline stdlib/Distributed/test/distributed_exec.jl]: `a27933`?

File: stdlib/Distributed/src/clusterserialize.jl (hunk 1 above)
Comment (JeffBezanson) [reply stdlib/Distributed/src/clusterserialize.jl]: That's a good point. I believe the value is unused though, so I'll add an explicit `return nothing`.

File: stdlib/Distributed/test/distributed_exec.jl (hunk 2 above)
Comment (JeffBezanson) [reply stdlib/Distributed/test/distributed_exec.jl]: Oops.

**Pull request merged**